sdk/core/azure-core/inc/http/bodyStream.hpp

```diff
@@ -0,0 +1,269 @@
+      }
+    }
+
+    static std::unique_ptr<std::vector<uint8_t>> ReadToEnd(Context& context, BodyStream& body)
+    {
+      constexpr int64_t chunkSize = 1024 * 8;
+      auto unique_buffer = std::make_unique<std::vector<uint8_t>>();
+
+      for (auto chunkNumber = 0;; chunkNumber++)
+      {
+        unique_buffer->resize((chunkNumber + 1) * chunkSize);
+        int64_t readBytes = ReadToCount(
+            context, body, unique_buffer->data() + (chunkNumber * chunkSize), chunkSize);
+
+        if (readBytes < chunkSize)
+        {
+          unique_buffer->resize((chunkNumber * chunkSize) + readBytes);
+          return unique_buffer;
+        }
+      }
+    }
+  };
+
+  class MemoryBodyStream : public BodyStream {
+  private:
+    const uint8_t* m_data;
+    int64_t m_length;
+    int64_t m_offset = 0;
+
+  public:
+    // Forbid constructor for rval so we don't end up storing dangling ptr
+    MemoryBodyStream(std::vector<uint8_t> const&&) = delete;
+
+    MemoryBodyStream(std::vector<uint8_t> const& buffer)
+        : MemoryBodyStream(buffer.data(), static_cast<uint64_t>(buffer.size()))
+    {
+    }
+
+    // cast as vector from ptr and length
+    explicit MemoryBodyStream(const uint8_t* data, int64_t length) : m_data(data), m_length(length)
+    {
+    }
+
+    int64_t Length() const override { return this->m_length; }
+
+    int64_t Read(Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      int64_t copy_length = std::min(count, (int64_t)this->m_length - this->m_offset);
+      // Copy what's left or just the count
+      std::memcpy(buffer, this->m_data + m_offset, (size_t)copy_length);
+      // move position
+      m_offset += copy_length;
+
+      return copy_length;
+    }
+
+    void Rewind() override { m_offset = 0; }
+
+    void Close() override {}
+  };
+
+  // Use for request with no body
+  class NullBodyStream : public Azure::Core::Http::BodyStream {
+  public:
+    explicit NullBodyStream() {}
+
+    ~NullBodyStream() override {}
+
+    int64_t Length() const override { return 0; }
+
+    void Rewind() override {}
+
+    int64_t Read(Azure::Core::Context& context, uint8_t* buffer, int64_t count) override
+    {
+      (void)context;
+      (void)buffer;
+      (void)count;
+      return 0;
+    };
+
+    void Close() override {}
+  };
+
+#ifdef Posix
+  class FileBodyStream : public BodyStream {
+  private:
+    // in mutable
+    int m_fd;
+    int64_t m_baseOffset;
+    int64_t m_length;
+    // mutable
+    int64_t m_offset;
+
+  public:
+    FileBodyStream(int fd, int64_t offset, int64_t length)
+        : m_fd(fd), m_baseOffset(offset), m_length(length), m_offset(0)
+    {
+    }
+
+    ~FileBodyStream() override {}
+
+    // Rewind seek back to 0
+    void Rewind() override { this->m_offset = 0; }
+
+    int64_t Read(Azure::Core::Context& context, uint8_t* buffer, int64_t count) override
+    {
+      context.ThrowIfCanceled();
+
+      auto result = pread(
```

Comment:
`pread()` may return `-1` on error.

Reply:
https://github.com/Azure/azure-sdk-for-cpp/issues/237